### ISSUE TITLE
Micrometer URI templating does not apply for unauthorized requests

### ISSUE THREAD

**Issue by ivansenic** [quarkusio/quarkus#24938]
### Describe the bug

In case a HTTP request ends up in `401 Unauthorized`, the Micrometer URI tag will not be templated, but include the full URI. If the request with the same URI is correctly authenticated, then the URI will be templated.

### Expected behavior

* URIs are correctly templated no matter what the response code is 

### Actual behavior

* does not work, at least with the `quarkus-resteasy-reactive`

### How to Reproduce?

Easy to reproduce with:

```
@Path("/hello")
public class GreetingResource {

    @GET
    @Path("something/{message}")
    @Produces(MediaType.TEXT_PLAIN)
    public Uni<RestResponse<String>> hello(@PathParam("message") String message) {
        return Uni.createFrom().item(RestResponse.ok(message));
    }
}
```

and:

```
%dev.quarkus.security.users.embedded.enabled=true
%dev.quarkus.security.users.embedded.plain-text=true
%dev.quarkus.security.users.embedded.users.scott=reader
%dev.quarkus.security.users.embedded.users.stuart=writer
%dev.quarkus.security.users.embedded.roles.scott=READER
%dev.quarkus.security.users.embedded.roles.stuart=READER,WRITER

quarkus.http.auth.permission.default.paths=/hello/*
quarkus.http.auth.permission.default.policy=authenticated
```

Results in the following URIs, which provide that `401` requests are not templated and `200` are:

```
http_server_requests_seconds_count{method="GET",outcome="CLIENT_ERROR",status="401",uri="/hello/something/my",} 3.0
http_server_requests_seconds_sum{method="GET",outcome="CLIENT_ERROR",status="401",uri="/hello/something/my",} 0.013654029
http_server_requests_seconds_count{method="GET",outcome="SUCCESS",status="200",uri="/hello/something/{message}",} 1.0
http_server_requests_seconds_sum{method="GET",outcome="SUCCESS",status="200",uri="/hello/something/{message}",} 0.029346689
http_server_requests_seconds_count{method="GET",outcome="CLIENT_ERROR",status="401",uri="/hello/something/asdsad",} 3.0
http_server_requests_seconds_sum{method="GET",outcome="CLIENT_ERROR",status="401",uri="/hello/something/asdsad",} 0.535294234
```

### Output of `uname -a` or `ver`

Linux ise-Precision-5550 5.13.0-39-generic #44~20.04.1-Ubuntu SMP Thu Mar 24 16:43:35 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Quarkus version or git rev

2.8.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537) Maven home: /home/ise/.m2/wrapper/dists/apache-maven-3.8.4-bin/52ccbt68d252mdldqsfsn03jlf/apache-maven-3.8.4 Java version: 17.0.2, vendor: Private Build, runtime: /usr/lib/jvm/java-17-openjdk-amd64 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.13.0-39-generic", arch: "amd64", family: "unix"

**Comment by ivansenic** [comment]
Anybody looking into this?

**Comment by ebullient** [comment]
I have a new person on the team now, and once he's done a few things to get oriented, I was going to have a try with this one with him. It is going to get into some nasty bits..

**Comment by ivansenic** [comment]
Using the match pattern does work and I am using this as the workaround at the moment.

@ebullient I also noticed that it's impossible to customize the response in case of `401 Unauthorized`. I followed the instructions here https://quarkus.io/guides/security-built-in-authentication#how-to-customize-authentication-exception-responses, however it does not end up in calling that mapper. Proactive auth on/off does not have any influence as well. Is thise somehow correlated? Should I open a separate issue?

**Comment by sberyozkin** [comment]
@ivansenic,

> @ebullient I also noticed that it's impossible to customize the response in case of 401 Unauthorized. I followed the instructions here https://quarkus.io/guides/security-built-in-authentication#how-to-customize-authentication-exception-responses, however it does not end up in calling that mapper. Proactive auth on/off does not have any influence as well. Is this somehow correlated? Should I open a separate issue?

Yes, please create a separate issue with a reproducer

**Comment by isacandrei** [comment]
I can confirm that this does not only affect status `401`, but also `409` and `RESET`. (Quarkus 2.7.5.Final)

**Comment by ebullient** [comment]
y.. all the 4xx shortcut the successful path that allows us to retrieve the template. There are some constraints that make build-time discovery of restful paths difficult, and getting the paths correctly affiliated with one of the constructed application paths when full invocation doesn't happen is non-trivial at best. This is why matchPattern exists. ;)

**Comment by isacandrei** [comment]
Thanks for the update. This is unfortunate though. Although matchPatterns solves the problem, if one would want to use these metrics across a wide set of microservices, each with 5-20 endpoints it would require considerable effort. Is there anything that can be done or I can help with to bring this home, or is it a technical limitation we cannot avoid?

**Comment by brunobat** [comment]
related to [RestClient errors statuscode are not shown in micrometer](https://github.com/quarkusio/quarkus/issues/25951)

**Comment by brunobat** [comment]
Took a look at this...
The only way I see to permanently fix this templating issue is to resolve the template and put it in the context, before the vertx authentication is executed... around here: io.quarkus.vertx.http.runtime.security.HttpAuthorizer:126
Vertx will delegate to resteasy classic only after that. No filters will help here.
This is non trivial, though.

**Comment by geoand** [comment]
As promised, it's [ugly](https://github.com/quarkusio/quarkus/pull/34702) but it works (I tested manually).

If we like this, I can add some tests and mark it ready for review

**Comment by brunobat** [comment]
Will this also work with the OpenTelemetry extension?
It suffers from the same condition.

**Comment by geoand** [comment]
Yes

**Comment by DaveK-KNIME** [comment]
I can reproduce the original issue when using `quarkus-resteasy` instead of `quarkus-rest` in Quarkus 3.15.4. Is this expected?

**Comment by geoand** [comment]
I am pretty sure we didn't do any work to make it work in Quarkus RESTEasy
